# 1. Index Categories

## 1.1 Equity Indices — China
- Shanghai Shenzhen CSI 300 Inde - SHSZ300 INDEX Index — Large-cap A-shares benchmark (top 300) for Mainland China.
- CSI Smallcap 500 Index - Shang - SH000905 INDEX Index — Mid/small-cap A-shares (500 constituents), complement to CSI 300.
- Shanghai Stock Exchange SSE 50 - SSE50 Index Index — 50 blue-chip A-shares on the Shanghai Stock Exchange.

## 1.2 Equity Futures
- Generic 1st 'ES' Future - ES1 Index Index — Front-month E-mini S&P 500 futures continuous contract.
- CSI 300 IDX FUTUR Dec25 - IFBZ5 Index Index — CSI 300 index futures, Dec 2025 contract.
- CSI500 IDX Future Dec25 - FFDZ5 Index Index — CSI 500 index futures, Dec 2025 contract.

## 1.3 Equity Options
- Deutsche Boerse AG German Stoc - DAX 10_17_25 P23000 INDEX Index — DAX index put option, 17 Oct 2025, strike 23,000.
- S&P 500 INDEX - SPX 12_19_25 P6000 Index Index — SPX index put option, 19 Dec 2025, strike 6,000.
- S&P 500 INDEX - SPX 12_20_30 P4300 Index Index — SPX index put option, 20 Dec 2030, strike 4,300.
- Nikkei 225 - NKY 10 P37250 INDEX Index — Nikkei 225 put option, Oct expiry, strike 37,250.
- Shanghai Stock Exchange SSE 50 - SSE50 9 C3000 INDEX Index — SSE 50 call option, Sep expiry, strike 3,000.
- S&P500 EMINI OPTN Dec25C  5000 - ESZ5C 5000 Index Index — E-mini S&P 500 call option, Dec 2025, strike 5,000.
- S&P500 EMINI OPTN Dec25P  4500 - ESZ5P 4500 Index Index — E-mini S&P 500 put option, Dec 2025, strike 4,500.

## 1.4 Volatility — Equity
- Cboe Volatility Index - VIX Index Index — 30-day implied volatility of the S&P 500.
- EURO STOXX 50 Volatility Index - V2X Index Index — 30-day implied volatility for EURO STOXX 50.
- Cboe Emerging Markets ETF EEM - VXEEM INDEX Index — Implied volatility index for EEM (Emerging Markets) ETF.

## 1.5 Volatility Options
- Cboe Volatility Index - VIX 12_17_25 C20 Index Index — VIX call option, 17 Dec 2025, strike 20.
- Cboe Volatility Index - VIX 12_17_25 P15 Index Index — VIX put option, 17 Dec 2025, strike 15.
- VSTOXX Opt        Sep25C    20 - FVSU5C 20 Index Index — VSTOXX call option, Sep 2025, strike 20.

## 1.6 Volatility — Rates
- ICE BofA MOVE Index - MOVE Index Index — Implied volatility of U.S. Treasury interest rates.

## 1.7 Rates & Yields
- Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index Curncy — U.S. Treasury 5-year nominal spot yield.
- US Generic Govt 10 Yr - USGG10YR Index Index — U.S. 10-year Treasury benchmark yield.

## 1.8 Money Markets & Policy
- ESTR - ESTRON INDEX Index — Euro short-term risk-free overnight rate.
- SHIBOR Fixing 3M - SHIF3M INDEX Index — Shanghai Interbank Offered Rate, 3-month tenor.
- Fed Funds Target Rate US - FDTR Index Index — Federal Funds target rate (upper bound).

## 1.9 Inflation-Linked & Breakevens
- BE 10 Year - USGGBE10 Index Index — U.S. 10-year breakeven inflation rate.
- Bloomberg US Govt Inflation-Li - BCIT1T Index Index — U.S. TIPS (govt inflation-linked) 1–10y index.
- Euro IL ex Italy 10+ Yrs - BDIF5Y  Index Index — Eurozone IL bonds ex-Italy, maturities 10+ years.

## 1.10 Fixed Income — Credit & Loans
- BBG EA Corp Avg OAS - LECPOAS INDEX Index — Euro area corporate bonds average OAS (credit spread).
- BBG PE HY Avg OAS - LP01OAS INDEX Index — Pan-Europe high-yield bonds average OAS (credit spread).
- iBoxx USD Liquid High Yield In - IBOXHY Index Index — USD liquid high-yield corporate index.
- iBoxx USD Liquid Investment Gr - IBOXIG Index Index — USD liquid investment-grade corporate index.
- Morningstar LSTA US Leveraged - SPBDLL Index Index — U.S. leveraged loan index (LSTA).

## 1.11 FX
- DOLLAR INDEX SPOT - DXY Index Index — U.S. Dollar Index versus major currencies (spot).

## 1.12 Crypto
- Bloomberg Galaxy Crypto Index - BGCI Index Index — Broad crypto market cap-weighted index.

## 1.13 Real Assets & Sectors
- FTSE NAREIT All Reits Total Re - FNARTR Index Index — U.S. REITs total return index.
- Dow Jones Brookfield Global In - DJBGICUT INDEX Index — Global listed infrastructure index.

## 1.14 ETFs & Mutual Funds
- Invesco S&P 500 Pure Growth ET - RPG US Equity Index — Invesco S&P 500 Pure Growth ETF.
- NYLI Hedge Multi-Strategy Trac - QAI US EQUITY Index — IQ Hedge Multi-Strategy ETF proxy for hedge funds.
- Vanguard 500 Index Fund - VFIAX US Equity US — Vanguard 500 Index mutual fund (S&P 500).
- Vanguard Balanced Index Fund - VBIAX US Equity US — Vanguard Balanced Index mutual fund (approx. 60/40).
- Vanguard Growth Index Fund - VIGAX US Equity US — Vanguard Growth Index mutual fund.
- Vanguard Value Index Fund - VVIAX US Equity US — Vanguard Value Index mutual fund.

## 1.15 Macro — Growth & Activity
- QoQ % Change Annualized - GDP CQOQ Index Index — U.S. real GDP quarter-over-quarter annualized growth.
- Manufacturing SA - CPMINDX INDEX Index — Manufacturing activity diffusion index (seasonally adjusted).
- Year % change - IP YOY Index Index — Industrial production, year-over-year percent change.
- Month % change - DGNOCHNG Index Index — Durable goods orders, monthly percent change.
- Monthly % Change - RSTAXMOM Index Index — Retail sales (ex-auto), month-over-month percent change.

## 1.16 Macro — Inflation (Prices)
- YoY % NSA - CPI YOY Index Index — CPI headline year-over-year (not seasonally adjusted).
- YoY % NSA - CPI XYOY Index Index — Core CPI (ex-food & energy) year-over-year (NSA).
- Monthly % Change - PCE CRCH Index Index — Core PCE price index month-over-month change.
- Goods YoY NSA - PPI YOY Index Index — Producer Price Index headline year-over-year (NSA).

## 1.17 Macro — Labor
- Net Change SA - NFP TCH Index Index — Nonfarm payrolls monthly net change (seasonally adjusted).
- Total SA - USURTOT Index Index — Unemployment rate, total (seasonally adjusted).

## 1.18 Macro — Housing
- Total authorized - NHSPATOT INDEX Index — Housing permits, total authorized (seasonally adjusted).

## 1.19 Macro — Money & Financial
- M2 (NSA) - M2NS Index Index — U.S. money stock M2 (not seasonally adjusted).
- US Financial Conditions FCON - BFCIUS INDEX Index — Bloomberg U.S. Financial Conditions Index.

## 1.20 Macro — Sentiment & Surprise
- Confidence - CONCCONF INDEX Index — Conference Board Consumer Confidence.
- Univ. of Michigan Sentiment - CONSSENT INDEX Index — University of Michigan Consumer Sentiment.
- Citi Economic Surprise - Unite - CESIUSD INDEX Index — Citigroup U.S. Economic Surprise Index.

## 1.21 Generic Futures (unspecified)
- Generic 1st 'CF' Future - CF1 Index Index — Generic front-month continuous future (symbol CF).
- Generic 1st 'GX' Future - GX1 Index Index — Generic front-month continuous future (symbol GX).
- Generic 1st 'Z ' Future - Z 1 Index Index — Generic front-month continuous future (symbol Z).



# 2. Index Analysis: Methodology, Outputs, and Interpretation

This notebook builds a clean, aligned, and feature-rich panel of market indices and derives interpretable diagnostics for downstream modelling and portfolio context. It follows a staged pipeline so you can inspect each transformation and its artefacts.

## 2.1 Objectives
- Ingest heterogeneous index workbooks and produce a single aligned panel.
- Denoise and transform prices into stationary series suitable for analysis.
- Engineer technical features and anomaly flags for monitoring and modelling.
- Extract regime signals and cross-sectional structure (ranks, PCA).
- Link the preprocessed data to macro and cross-asset context.

## 2.2 Data Inputs & Assumptions
- Input directory: `INDEX/*.xlsx` containing time series with a date column and one or more value columns.
- A primary value column is auto-selected from common aliases (e.g., adjusted close, close).
- Data are aligned to business-day frequency; gaps can remain when sources are sparse.
- Where appropriate, forward/backward fill and interpolation are applied for continuity.

## 2.3 Pipeline Map (What each stage does)
- Stage 0 — Helpers: Reusable utilities for filling, filtering, features, regimes.
- Stage 1 — Cleaning & Alignment: Canonicalise frames, choose primary column, align to business days, and summarise coverage.
- Stages 2–3 — Noise Reduction & Stationarity: Moving-average and Fourier low-pass smoothing; transform to log-returns or differences; scale by rolling volatility.
- Stages 4–5 — Features & Anomalies: Rolling moments, RSI/MACD/Bollinger Bands, higher moments, winsorised returns, z-score anomaly flags.
- Stages 6–8 — Selection, Regimes, Cross-Section: PCA exposures, optional mutual-information ranking vs a target (e.g., SPX), regime detection, dynamic correlations, cross-sectional ranks.
- Stages 9–10 — Augmentation & Bundle: Optional block bootstrap for stress testing; assemble everything into `nexus_index_bundle` for easy reuse.
- Market Dynamics — Interpretation: Macro linkage, lead–lag, regime diagnostics, and behavioural proxies.

## 2.4 Key Artefacts (Where to look)
- `aligned_panel` (Stage 1): Unified DataFrame of primary value series indexed by date.
- `stage1_summary` (Stage 1): Per-series start/end, observation count, and data `coverage`.
- `stage2_smoothed` (Stage 2): Smoothed versions of the aligned panel (moving average and Fourier low-pass).
- `stationary_bundle` (Stage 3): `{ 'returns', 'diff', 'rolling_vol_20', 'vol_scaled' }`.
- `stage4_feature_panel` (Stage 4): Technical indicators per series (may be a MultiIndex on columns).
- `anomaly_flags`, `higher_moments`, `winsorized_returns` (Stage 5).
- `pca_features`, `mutual_information_rank` (Stage 6).
- `regimes`, `regime_scaled_returns`, `dynamic_corr_to_spx` (Stage 7).
- `cross_sectional_rank` (Stage 8).
- `bootstrap_samples`, `bootstrap_summary` (Stage 9).
- `nexus_index_bundle` (Stage 10): A dictionary exposing all of the above for downstream use.

## 2.5 How To Read The Results
- Coverage: In `stage1_summary`, prefer series with higher `coverage` for robust analysis; flag sparse series for manual review.
- Stationarity: Confirm `returns` and/or `diff` are mean-reverting enough for your modelling horizon; check `rolling_vol_20` for stability.
- Features: Inspect RSI, MACD, Bollinger Bands alongside price to validate signal behaviour (e.g., RSI extremes).
- Anomalies: Large z-scores highlight jumps/spikes; compare outcomes in `winsorized_returns` to control for outliers.
- PCA/Selection: Use `pca_features` to understand common drivers; `mutual_information_rank` can highlight series most informative to a chosen target.
- Regimes: Validate `regimes` (bull/bear/stagnant) against known episodes; use `regime_scaled_returns` for regime-aware normalisation.
- Cross-Asset: Check `dynamic_corr_to_spx` and `lead_lag_table` for propagation chains.
- Macro: Use `macro_beta_table` and `macro_correlations` (built in Market Dynamics) to sanity-check factor directionality and strength.

## 2.6 Limitations & Notes
- Smoothing and differencing choices can affect signal timing; verify against your use case.
- Mutual information ranking falls back to correlation if scikit-learn is unavailable.
- Regime detection uses simple moving-average comparisons; consider more advanced state models if needed.
- Bootstrap is illustrative and path-length matched; adapt block sizes to your series memory.


In [110]:
from pathlib import Path
import pandas as pd

index_folder = Path('INDEX')
xlsx_files = sorted(index_folder.glob('*.xlsx'))

if not xlsx_files:
    raise FileNotFoundError(f'No .xlsx files found in {index_folder.resolve()}')

index_data = {path.stem: pd.read_excel(path) for path in xlsx_files}
print(f'Loaded {len(index_data)} files from {index_folder.resolve()}')
print('First few entries:', list(index_data.keys())[:5])


Loaded 64 files from /Users/kechengshi/Documents/GitHub/Nexus-data-preprocessing/INDEX
First few entries: ['BBG EA Corp Avg OAS - LECPOAS INDEX Index', 'BBG PE HY Avg OAS - LP01OAS INDEX Index', 'BE 10 Year - USGGBE10 Index Index', 'Bloomberg Galaxy Crypto Index - BGCI Index Index', 'Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index Curncy']


# 3. Stage 0 - Helpers
Import the reusable utility functions mirrored from `Nexus Data Preprocessing.ipynb` so every downstream stage can reuse the Nexus transformations.


In [111]:
import numpy as np
from typing import Dict, List, Optional, Sequence


#
# - Fill NaNs forward then backward for specified columns.
# - Returns a copy to avoid mutating input frame.
def forward_backward_fill(frame: pd.DataFrame, columns: Optional[Sequence[str]] = None) -> pd.DataFrame:
    columns = list(columns) if columns is not None else list(frame.columns)
    out = frame.copy()
    out[columns] = out[columns].ffill().bfill()
    return out


#
# - Resample by `freq` using ffill/bfill or interpolation.
# - Keeps `date_col` as a column after reset_index().
def align_to_frequency(df: pd.DataFrame, date_col: str, freq: str, method: str = "ffill") -> pd.DataFrame:
    out = df.set_index(date_col).sort_index()
    if method == "ffill":
        out = out.resample(freq).ffill()
    elif method == "bfill":
        out = out.resample(freq).bfill()
    else:
        out = out.resample(freq).interpolate(method=method)
    return out.reset_index()


#
# - Simple rolling mean denoiser with min_periods=1.
def moving_average_filter(series: pd.Series, window: int = 5) -> pd.Series:
    return series.rolling(window=window, min_periods=1, center=False).mean()


#
# - Zero high-frequency FFT coefficients beyond cutoff_ratio.
# - Inverse transform and preserve original index length.
def fourier_lowpass_filter(series: pd.Series, cutoff_ratio: float = 0.1) -> pd.Series:
    values = series.to_numpy()
    n = values.shape[0]
    if n == 0:
        return series.copy()
    fft = np.fft.rfft(values)
    cutoff = int(len(fft) * cutoff_ratio)
    filtered_fft = np.zeros_like(fft)
    filtered_fft[:max(cutoff, 1)] = fft[:max(cutoff, 1)]
    filtered = np.fft.irfft(filtered_fft, n=n)
    return pd.Series(filtered, index=series.index)


#
# - Standardize to mean 0, std 1; guard for std==0.
def zscore_normalize(series: pd.Series) -> pd.Series:
    mean = series.mean()
    std = series.std()
    if std == 0:
        return series - mean
    return (series - mean) / std


#
# - Log difference of levels; first observation is NaN.
# - Ensure inputs are positive/non-zero upstream to avoid infs.
def compute_log_returns(series: pd.Series) -> pd.Series:
    return np.log(series).diff()


#
# - Scale returns by provided volatility; zero vol -> NaN.
def volatility_scale(returns: pd.Series, vol: pd.Series) -> pd.Series:
    return returns / vol.replace(0.0, np.nan)


#
# - Clip series at given lower/upper quantiles to reduce outliers.
def winsorize(series: pd.Series, lower_q: float = 0.01, upper_q: float = 0.99) -> pd.Series:
    lo, hi = series.quantile([lower_q, upper_q])
    return series.clip(lower=lo, upper=hi)


#
# - Return DataFrame of rolling mean and std.
def rolling_window_features(series: pd.Series, window: int = 20) -> pd.DataFrame:
    roll_mean = series.rolling(window=window, min_periods=window).mean()
    roll_std = series.rolling(window=window, min_periods=window).std()
    return pd.DataFrame({"rolling_mean": roll_mean, "rolling_std": roll_std})


#
# - EWM-based RSI with default 14 periods (0–100).
def compute_rsi(prices: pd.Series, window: int = 14) -> pd.Series:
    delta = prices.diff()
    gain = delta.clip(lower=0).ewm(alpha=1 / window, adjust=False, min_periods=window).mean()
    loss = -delta.clip(upper=0).ewm(alpha=1 / window, adjust=False, min_periods=window).mean()
    rs = gain / loss.replace(0, np.nan)
    rsi = 100 - (100 / (1 + rs))
    return rsi


#
# - EMA(12/26) MACD, 9-period signal, and histogram.
def compute_macd(prices: pd.Series, span_fast: int = 12, span_slow: int = 26, span_signal: int = 9) -> pd.DataFrame:
    ema_fast = prices.ewm(span=span_fast, adjust=False).mean()
    ema_slow = prices.ewm(span=span_slow, adjust=False).mean()
    macd_line = ema_fast - ema_slow
    signal = macd_line.ewm(span=span_signal, adjust=False).mean()
    hist = macd_line - signal
    return pd.DataFrame({"macd": macd_line, "macd_signal": signal, "macd_hist": hist})


#
# - Rolling mean ± n_std*std; returns mid/upper/lower bands.
def compute_bbands(prices: pd.Series, window: int = 20, n_std: float = 2.0) -> pd.DataFrame:
    ma = prices.rolling(window=window, min_periods=window).mean()
    sd = prices.rolling(window=window, min_periods=window).std()
    upper = ma + n_std * sd
    lower = ma - n_std * sd
    return pd.DataFrame({"bb_mid": ma, "bb_upper": upper, "bb_lower": lower})


#
# - Rolling skewness and kurtosis over window (default 60).
def rolling_higher_moments(returns: pd.Series, window: int = 60) -> pd.DataFrame:
    skew = returns.rolling(window=window, min_periods=window).skew()
    kurt = returns.rolling(window=window, min_periods=window).kurt()
    return pd.DataFrame({"rolling_skew": skew, "rolling_kurt": kurt})


#
# - MA crossover labels: bull (fast>slow), bear (fast<slow), else stagnant.
def detect_market_regimes(series: pd.Series, fast_window: int = 20, slow_window: int = 100) -> pd.Series:
    fast_ma = series.rolling(window=fast_window, min_periods=1).mean()
    slow_ma = series.rolling(window=slow_window, min_periods=1).mean()
    regime = pd.Series("stagnant", index=series.index)
    regime = regime.mask(fast_ma > slow_ma, "bull")
    regime = regime.mask(fast_ma < slow_ma, "bear")
    return regime


#
# - Z-score within contiguous regime segments; preserves index.
def regime_specific_normalize(series: pd.Series, regimes: pd.Series) -> pd.Series:
    series = series.copy()
    regimes = regimes.reindex(series.index)
    out = series.copy()
    for regime, idx in regimes.groupby(regimes).groups.items():
        labels = list(idx)
        normalized = zscore_normalize(series.loc[labels])
        out.loc[labels] = normalized
    return out


#
# - Binary flag where |z| exceeds threshold (default 3.0).
def detect_anomalies_zscore(series: pd.Series, z_thresh: float = 3.0) -> pd.Series:
    z = zscore_normalize(series)
    return (np.abs(z) > z_thresh).astype(int)


#
# - Rolling correlation with min_periods=window.
def dynamic_correlation(series_a: pd.Series, series_b: pd.Series, window: int = 60) -> pd.Series:
    return series_a.rolling(window=window, min_periods=window).corr(series_b)


#
# - Center features, eigendecompose covariance, return top PCs.
def principal_component_features(features: pd.DataFrame, n_components: int = 3) -> pd.DataFrame:
    centered = features - features.mean()
    cov = np.cov(centered.fillna(0.0).to_numpy().T)
    eigvals, eigvecs = np.linalg.eigh(cov)
    order = np.argsort(eigvals)[::-1][:n_components]
    pcs = centered.to_numpy() @ eigvecs[:, order]
    columns = [f"pc{i+1}" for i in range(len(order))]
    return pd.DataFrame(pcs, index=features.index, columns=columns)


#
# - Rank features by MI vs target (fallback: |corr|).
def mutual_information_ranking(features: pd.DataFrame, target: pd.Series) -> pd.Series:
    try:
        from sklearn.feature_selection import mutual_info_regression  # type: ignore
    except Exception:
        corr = features.corrwith(target)
        return corr.abs().sort_values(ascending=False)
    mi = mutual_info_regression(features.fillna(0.0), target.fillna(0.0))
    return pd.Series(mi, index=features.columns).sort_values(ascending=False)


#
# - Per-date ranks scaled to (0,1) via average method.
def cross_sectional_rank_normalize(df: pd.DataFrame, date_col: str, value_col: str) -> pd.Series:
    ranks = df.groupby(date_col)[value_col].rank(method="average")
    group_sizes = df.groupby(date_col)[value_col].transform("count")
    return (ranks - 0.5) / group_sizes


#
# - Sample fixed-size blocks to preserve autocorrelation structure.
def block_bootstrap(series: pd.Series, block_size: int = 5, n_samples: int = 100, random_state: Optional[int] = None) -> np.ndarray:
    if block_size <= 0:
        raise ValueError('block_size must be positive')
    rng = np.random.default_rng(random_state)
    values = series.dropna().to_numpy()
    if len(values) == 0:
        return np.empty((n_samples, 0))
    blocks = [values[i:i + block_size] for i in range(0, len(values), block_size)]
    if not blocks:
        return np.empty((n_samples, 0))
    n_blocks_needed = max(int(np.ceil(len(values) / block_size)), 1)
    bootstrapped = np.empty((n_samples, len(values)))
    for sample_idx in range(n_samples):
        picks = rng.integers(0, len(blocks), size=n_blocks_needed)
        stitched = np.concatenate([blocks[idx] for idx in picks])
        while stitched.shape[0] < len(values):
            extra_idx = int(rng.integers(0, len(blocks)))
            stitched = np.concatenate([stitched, blocks[extra_idx]])
        bootstrapped[sample_idx, :] = stitched[:len(values)]
    return bootstrapped



# 4. Stage 1 - Data Cleaning and Alignment
Canonicalise each raw workbook, resample to business days, pick a primary value column, and build the aligned panel plus quick coverage summary.


In [112]:
from collections import OrderedDict

preferred_value_aliases = [
    "adj_close",
    "adjusted close",
    "close",
    "px_last",
    "last price",
    "price",
    "value",
    "index",
]


#
# - Normalize headers, parse dates, cast numerics, fill, deduplicate.
def canonicalize_index_frame(df: pd.DataFrame) -> pd.DataFrame:
    frame = df.copy()
    frame = frame.dropna(how="all")
    frame.columns = [
        (str(col).strip() if not str(col).startswith("Unnamed") else f"col_{i}")
        for i, col in enumerate(frame.columns)
    ]
    if frame.empty:
        return frame
    first_col = frame.columns[0]
    frame = frame.rename(columns={first_col: "date"})
    if isinstance(frame["date"].iloc[0], str) and "date" in frame["date"].iloc[0].lower():
        frame = frame.iloc[1:].reset_index(drop=True)
    frame["date"] = pd.to_datetime(frame["date"], errors="coerce")
    frame = frame.dropna(subset=["date"])
    numeric_cols = [col for col in frame.columns if col != "date"]
    for col in numeric_cols:
        frame[col] = pd.to_numeric(frame[col], errors="coerce")
    numeric_cols = [col for col in numeric_cols if not frame[col].isna().all()]
    frame = frame[["date"] + numeric_cols]
    if numeric_cols:
        frame = forward_backward_fill(frame.sort_values("date"), numeric_cols)
    frame = frame.drop_duplicates(subset=["date"]).reset_index(drop=True)
    return frame


#
# - Choose value column by common aliases; fallback to first column.
def pick_primary_column(columns: Sequence[str]) -> str:
    normalized = {col: col.lower().strip() for col in columns}
    for alias in preferred_value_aliases:
        for col in columns:
            if alias in normalized[col]:
                return col
    return columns[0]


stage1_frames: Dict[str, pd.DataFrame] = {}
primary_columns: Dict[str, str] = {}
for name, raw in index_data.items():
    tidy = canonicalize_index_frame(raw)
    if tidy.empty or tidy.shape[1] <= 1:
        continue
    aligned = align_to_frequency(tidy, "date", "B", method="ffill")
    value_cols = [col for col in aligned.columns if col != "date"]
    if not value_cols:
        continue
    primary_columns[name] = pick_primary_column(value_cols)
    stage1_frames[name] = aligned

aligned_panel = pd.DataFrame({
    name: frame.set_index("date")[primary_columns[name]]
    for name, frame in stage1_frames.items()
}).sort_index()
aligned_panel = aligned_panel.loc[:, ~aligned_panel.columns.duplicated()]
aligned_panel = aligned_panel.dropna(how="all")

#
# - Optional: mitigate sparse tail by modest ffill and coverage trim.
#   This addresses trailing all-NaN rows seen in Stage 2/3 outputs
#   when only a few series extend to the latest dates.
MIN_DAILY_COVERAGE_TAIL = 0.10  # keep dates with >=10% non-NaN series
FFILL_TAIL_LIMIT = 3            # forward-fill up to 3 business days

# Small ffill to cover short holiday gaps without extending stale levels too far.
aligned_panel = aligned_panel.ffill(limit=FFILL_TAIL_LIMIT)

# Trim any remaining trailing dates with very low cross-series coverage.
if not aligned_panel.empty:
    daily_cov = 1.0 - aligned_panel.isna().mean(axis=1)
    if (daily_cov >= MIN_DAILY_COVERAGE_TAIL).any():
        last_good = daily_cov.index[(daily_cov >= MIN_DAILY_COVERAGE_TAIL)].max()
        aligned_panel = aligned_panel.loc[:last_good]

stage1_summary = pd.DataFrame([
    {
        "series": name,
        "primary_column": primary_columns.get(name),
        "start": frame["date"].min(),
        "end": frame["date"].max(),
        "observations": frame["date"].nunique(),
    }
    for name, frame in stage1_frames.items()
    if name in primary_columns
])
if not aligned_panel.empty:
    coverage = 1.0 - aligned_panel.isna().mean()
    stage1_summary["coverage"] = stage1_summary["series"].map(coverage)
else:
    stage1_summary["coverage"] = np.nan

nexus_index_stages = {
    "stage1": {
        "frames": stage1_frames,
        "panel": aligned_panel,
        "primary_columns": primary_columns,
        "summary": stage1_summary,
    }
}

stage1_summary.head(10)


series primary_column  \
0          BBG EA Corp Avg OAS - LECPOAS INDEX Index          col_1   
1            BBG PE HY Avg OAS - LP01OAS INDEX Index          col_4   
2                  BE 10 Year - USGGBE10 Index Index          col_1   
3   Bloomberg Galaxy Crypto Index - BGCI Index Index          col_1   
4  Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index...          col_1   
5  Bloomberg US Govt Inflation-Li - BCIT1T Index ...          col_1   
6        CSI 300 IDX FUTUR Dec25 - IFBZ5 Index Index          col_1   
7  CSI Smallcap 500 Index - Shang - SH000905 INDE...          col_1   
8        CSI500 IDX Future Dec25 - FFDZ5 Index Index          col_1   
9  Cboe Emerging Markets ETF EEM - VXEEM INDEX Index          col_1   

       start        end  observations  coverage  
0 2010-01-04 2025-09-16          4097  0.386719  
1 2010-01-04 2025-09-16          4097  0.386719  
2 2005-01-03 2025-09-16          5402  0.509809  
3 2018-01-02 2025-09-16          2011  0.189964  
4 2020-10-28 2025-09-16          1275  0.120543  
5 2005-01-03 2025-09-16          5402  0.509809  
6 2015-04-20 2025-09-16          2717  0.256555  
7 2015-01-05 2025-09-16          2792  0.263630  
8 2015-04-16 2025-09-16          2719  0.256744  
9 2011-03-16 2025-09-16          3785  0.357291

# 5. Stages 2-3 - Noise Reduction and Stationarity
Apply smoothing filters, log-return or difference transforms, and volatility scaling to prepare stationary series for feature work.


In [113]:
if aligned_panel.empty:
    raise ValueError("Stage 1 produced no aligned panel to process.")


#
# - Interpolate small gaps, apply Fourier low-pass, restore NaNs.
def _apply_lowpass(series: pd.Series) -> pd.Series:
    filled = series.interpolate(limit_direction="both")
    filtered = fourier_lowpass_filter(filled)
    filtered[series.isna()] = np.nan
    return filtered


moving_average_panel = aligned_panel.apply(moving_average_filter)
lowpass_panel = aligned_panel.apply(_apply_lowpass)

stage2_smoothed = pd.concat(
    {
        "moving_average": moving_average_panel,
        "fourier_lowpass": lowpass_panel,
    },
    axis=1,
)

returns = aligned_panel.apply(lambda s: compute_log_returns(s.replace(0.0, np.nan)))
returns = returns.replace([np.inf, -np.inf], np.nan)
diffs = aligned_panel.diff()
rolling_vol_20 = returns.rolling(window=20, min_periods=20).std()
vol_scaled_returns = pd.DataFrame({
    col: volatility_scale(returns[col], rolling_vol_20[col])
    for col in returns.columns
})

stationary_bundle = {
    "returns": returns,
    "diff": diffs,
    "rolling_vol_20": rolling_vol_20,
    "vol_scaled": vol_scaled_returns,
}

nexus_index_stages.update({
    "stage2": {"smoothed": stage2_smoothed},
    "stage3": stationary_bundle,
})

stage2_smoothed.tail()


/opt/anaconda3/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: invalid value encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


moving_average  \
           BBG EA Corp Avg OAS - LECPOAS INDEX Index   
date                                                   
2025-09-15                                     0.798   
2025-09-16                                     0.796   
2025-09-17                                     0.794   
2025-09-18                                     0.792   
2025-09-19                                     0.790   

                                                    \
           BBG PE HY Avg OAS - LP01OAS INDEX Index   
date                                                 
2025-09-15                                   2.920   
2025-09-16                                   2.898   
2025-09-17                                   2.880   
2025-09-18                                   2.868   
2025-09-19                                   2.862   

                                              \
           BE 10 Year - USGGBE10 Index Index   
date                                           
2025-09-15                           2.36374   
2025-09-16                           2.36576   
2025-09-17                           2.36722   
2025-09-18                           2.37048   
2025-09-19                           2.37172   

                                                             \
           Bloomberg Galaxy Crypto Index - BGCI Index Index   
date                                                          
2025-09-15                                         3827.786   
2025-09-16                                         3864.138   
2025-09-17                                         3890.246   
2025-09-18                                         3903.934   
2025-09-19                                         3890.800   

                                                                  \
           Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index Curncy   
date                                                               
2025-09-15                                           85.17362      
2025-09-16                                           85.18408      
2025-09-17                                           85.18046      
2025-09-18                                           85.17560      
2025-09-19                                           85.20212      

                                                                \
           Bloomberg US Govt Inflation-Li - BCIT1T Index Index   
date                                                             
2025-09-15                                          376.16454    
2025-09-16                                          376.44008    
2025-09-17                                          376.61910    
2025-09-18                                          376.67104    
2025-09-19                                          376.81540    

                                                        \
           CSI 300 IDX FUTUR Dec25 - IFBZ5 Index Index   
date                                                     
2025-09-15                                     4449.68   
2025-09-16                                     4468.52   
2025-09-17                                     4490.56   
2025-09-18                                     4513.36   
2025-09-19                                     4509.36   

                                                                  \
           CSI Smallcap 500 Index - Shang - SH000905 INDEX Index   
date                                                               
2025-09-15                                          7016.2750      
2025-09-16                                          7054.7252      
2025-09-17                                          7102.9404      
2025-09-18                                          7148.5172      
2025-09-19                                          7155.8328      

                                                        \
           CSI500 IDX Future Dec25 - FFDZ5 Index Index   
date                                                     
2025-09-

# 6. Stages 4-5 - Feature Engineering and Anomalies
Generate technical indicators, rolling moments, winsorised returns, and z-score anomaly flags for each index.


In [114]:
winsorized_returns = vol_scaled_returns.apply(lambda s: winsorize(s))
higher_moments_panel = pd.concat(
    {
        col: rolling_higher_moments(returns[col]).rename(columns=lambda c: f"{c}")
        for col in returns.columns
    },
    axis=1,
)

technical_features = {}
for name in aligned_panel.columns:
    features = [
        rolling_window_features(aligned_panel[name]).add_prefix("roll_"),
        compute_rsi(aligned_panel[name]).to_frame("rsi14"),
        compute_macd(aligned_panel[name]).rename(columns=lambda c: f"macd_{c}"),
        compute_bbands(aligned_panel[name]).rename(columns=lambda c: f"bb_{c}"),
        rolling_higher_moments(returns[name]).rename(columns=lambda c: f"moment_{c}"),
    ]
    technical_features[name] = pd.concat(features, axis=1)

stage4_feature_panel = pd.concat(technical_features, axis=1)
anomaly_flags = winsorized_returns.apply(lambda s: detect_anomalies_zscore(s)).astype(int)

nexus_index_stages.update({
    "stage4": {"features": stage4_feature_panel},
    "stage5": {
        "winsorized_returns": winsorized_returns,
        "higher_moments": higher_moments_panel,
        "anomaly_flags": anomaly_flags,
    },
})

anomaly_flags.tail()


BBG EA Corp Avg OAS - LECPOAS INDEX Index  \
date                                                    
2025-09-15                                          0   
2025-09-16                                          0   
2025-09-17                                          0   
2025-09-18                                          0   
2025-09-19                                          0   

            BBG PE HY Avg OAS - LP01OAS INDEX Index  \
date                                                  
2025-09-15                                        0   
2025-09-16                                        0   
2025-09-17                                        0   
2025-09-18                                        0   
2025-09-19                                        0   

            BE 10 Year - USGGBE10 Index Index  \
date                                            
2025-09-15                                  0   
2025-09-16                                  0   
2025-09-17                                  0   
2025-09-18                                  0   
2025-09-19                                  0   

            Bloomberg Galaxy Crypto Index - BGCI Index Index  \
date                                                           
2025-09-15                                                 0   
2025-09-16                                                 0   
2025-09-17                                                 0   
2025-09-18                                                 0   
2025-09-19                                                 0   

            Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index Curncy  \
date                                                                
2025-09-15                                                  0       
2025-09-16                                                  0       
2025-09-17                                                  0       
2025-09-18                                                  0       
2025-09-19                                                  0       

            Bloomberg US Govt Inflation-Li - BCIT1T Index Index  \
date                                                              
2025-09-15                                                  0     
2025-09-16                                                  0     
2025-09-17                                                  0     
2025-09-18                                                  0     
2025-09-19                                                  0     

            CSI 300 IDX FUTUR Dec25 - IFBZ5 Index Index  \
date                                                      
2025-09-15                                            0   
2025-09-16                                            0   
2025-09-17                                            0   
2025-09-18                                            0   
2025-09-19                                            0   

            CSI Smallcap 500 Index - Shang - SH000905 INDEX Index  \
date                                                                
2025-09-15                                                  0       
2025-09-16                                                  0       
2025-09-17                                                  0       
2025-09-18                                                  0       
2025-09-19                                                  0       

            CSI500 IDX Future Dec25 - FFDZ5 Index Index  \
date                                                      
2025-09-15                                            0   
2025-09-16                                            0   
2025-09-17                                            0   
2025-09-18                                            0   
2025-09-19                                            0   

            Cboe Emerging Markets ETF EEM - VXEEM INDEX Index  ...  \
date                                                           ...   
2025-09-15                 

# 7. Stages 6-8 - Selection, Regimes, Cross-Section
Derive PCA exposures, optional mutual-information ranking versus SPX, regime-aware scaling, dynamic correlations, and cross-sectional ranks.


In [115]:
valid_returns = returns.dropna(how="all")
if not valid_returns.empty:
    filled_returns = valid_returns.fillna(method="ffill").fillna(method="bfill")
    n_components = min(5, max(1, filled_returns.shape[1]))
    pca_features = principal_component_features(filled_returns, n_components=n_components)
else:
    pca_features = pd.DataFrame()

spx_candidate = next((col for col in aligned_panel.columns if "spx" in col.lower() or "s&p" in col.lower()), None)
if spx_candidate and spx_candidate in returns.columns and returns.shape[1] > 1:
    features_for_mi = returns.drop(columns=[spx_candidate])
    mi_rank = mutual_information_ranking(features_for_mi, returns[spx_candidate])
else:
    mi_rank = None

regimes = pd.DataFrame({
    name: detect_market_regimes(aligned_panel[name])
    for name in aligned_panel.columns
})
regime_scaled_returns = pd.DataFrame({
    name: regime_specific_normalize(returns[name], regimes[name])
    for name in returns.columns
})

if spx_candidate and spx_candidate in returns.columns:
    dyn_corr = pd.DataFrame({
        name: dynamic_correlation(returns[name], returns[spx_candidate])
        for name in returns.columns
        if name != spx_candidate
    })
else:
    dyn_corr = pd.DataFrame()

stacked_returns = returns.stack().rename("return").reset_index()
stacked_returns.columns = ["date", "series", "return"]
stacked_returns["cs_rank"] = cross_sectional_rank_normalize(stacked_returns, "date", "return")
cross_sectional_panel = stacked_returns.pivot(index="date", columns="series", values="cs_rank")

nexus_index_stages.update({
    "stage6": {
        "pca": pca_features,
        "mutual_information_rank": mi_rank,
    },
    "stage7": {
        "regimes": regimes,
        "regime_scaled_returns": regime_scaled_returns,
        "dynamic_corr_to_spx": dyn_corr,
    },
    "stage8": {
        "cross_sectional_rank": cross_sectional_panel,
    },
})

pca_features.head()


/var/folders/03/s5l6v1614q77wz_4t437rdpr0000gn/T/ipykernel_10196/2771946356.py:3: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  filled_returns = valid_returns.fillna(method="ffill").fillna(method="bfill")


pc1       pc2      pc3       pc4       pc5
date                                                       
1985-02-01  0.056299 -0.115007 -0.07826  0.002915 -0.001341
1985-02-04  0.056299 -0.115007 -0.07826  0.002915 -0.001341
1985-02-05  0.056299 -0.115007 -0.07826  0.002915 -0.001341
1985-02-06  0.056299 -0.115007 -0.07826  0.002915 -0.001341
1985-02-07  0.056299 -0.115007 -0.07826  0.002915 -0.001341

# 8. Stages 9-10 - Augmentation and Pipeline Bundle
Run block-bootstrap augmentation (if SPX is present), and store every artifact in the `nexus_index_bundle` summary table for modelling pipelines.


In [116]:
bootstrap_samples: Dict[str, np.ndarray] = {}
if spx_candidate and spx_candidate in returns.columns:
    bootstrap_samples[spx_candidate] = block_bootstrap(returns[spx_candidate], block_size=5, n_samples=200, random_state=42)
bootstrap_summary = {
    name: {
        "mean_mean": samples.mean(axis=1).mean(),
        "mean_std": samples.mean(axis=1).std(),
        "path_std_mean": samples.std(axis=1).mean(),
    }
    for name, samples in bootstrap_samples.items()
}

nexus_index_bundle = {
    "aligned_panel": aligned_panel,
    "smoothed": stage2_smoothed,
    "stationary": stationary_bundle,
    "features": stage4_feature_panel,
    "anomalies": anomaly_flags,
    "pca": pca_features,
    "regimes": regimes,
    "regime_scaled_returns": regime_scaled_returns,
    "dynamic_corr_to_spx": dyn_corr,
    "cross_sectional_rank": cross_sectional_panel,
    "bootstrap_samples": bootstrap_samples,
    "bootstrap_summary": bootstrap_summary,
}

nexus_index_stages.update({
    "stage9": {
        "bootstrap_samples": bootstrap_samples,
        "bootstrap_summary": bootstrap_summary,
    },
    "stage10": {
        "bundle": nexus_index_bundle,
    },
})

summary_table = pd.DataFrame({
    "stage": list(nexus_index_stages.keys()),
    "artifacts": [", ".join(sorted(map(str, section.keys()))) for section in nexus_index_stages.values()],
})
summary_table


stage                                          artifacts
0   stage1            frames, panel, primary_columns, summary
1   stage2                                           smoothed
2   stage3          diff, returns, rolling_vol_20, vol_scaled
3   stage4                                           features
4   stage5  anomaly_flags, higher_moments, winsorized_returns
5   stage6                       mutual_information_rank, pca
6   stage7  dynamic_corr_to_spx, regime_scaled_returns, re...
7   stage8                               cross_sectional_rank
8   stage9               bootstrap_samples, bootstrap_summary
9  stage10                                             bundle

# 9. Market Dynamics Analysis
Link the staged preprocessing outputs to macro drivers, cross-asset influence, regime diagnostics, and behavioural signals for portfolio context.


In [117]:
if 'nexus_index_bundle' not in globals():
    raise RuntimeError("Run the preprocessing stages to create 'nexus_index_bundle' before executing the market dynamics analysis.")

bundle = nexus_index_bundle
aligned_panel = bundle['aligned_panel']
returns = bundle['stationary']['returns']
diffs = bundle['stationary']['diff']

macro_indicator_keywords = [
    'gdp', 'pmi', 'cpi', 'ppi', 'pce', 'yoy', 'qoq', 'mom', 'monthly', 'net', 'nfp',
    'unemployment', 'sentiment', 'confidence', 'surprise', 'm2', 'industrial', 'manufacturing',
    'housing', 'usurtot', 'nhspatot', 'cesi', 'rstax', 'dgnochng', 'authorized',
    'financial', 'conditions', 'goods', 'production', 'retail'
]
asset_rules = [
    ('Crypto', ['crypto', 'bgci']),
    ('Volatility', ['vix', 'volatility', 'v2x', 'fvsu5c', 'move']),
    ('FX', ['dxy', 'dollar', 'curncy']),
    ('Commodities', ['cf1', ' z', 'commodity', 'metals']),
    ('Credit', ['oas', 'iboxhy', 'iboxig', 'ibox', 'credit', 'leveraged', 'spbdll']),
    ('Rates', ['yield', 'rate', 'usgg10yr', 'usgg', 'shibor', 'estr', 'fdtr', 'bdif5y', 'btsi', 'bcit1t', 'breakeven', 'inflation-li', 'usggbe10']),
    ('Alternative', ['qai', 'multi', 'balanced']),
    ('Equity', ['equity', 'spx', 'sp500', 's&p', 'nikkei', 'nky', 'csi', 'csi500', 'ifbz5', 'shsz300', 'sse50', 'dax', 'djbgicut', 'fnartr', 'es1', 'esz5c', 'esz5p', 'esz', 'emini', 'rpg', 'vfiax', 'vbi', 'vvi', 'vigax', 'gx1', 'eem', 'reit'])
]
macro_keywords = {
    'Growth': ['gdp', 'industrial', 'nfp', 'unemployment', 'pmi', 'manufacturing', 'confidence', 'sentiment'],
    'Inflation': ['cpi', 'inflation', 'ppi', 'pce', 'deflator'],
    'Liquidity': ['m2', 'credit', 'financing'],
    'Policy Rates': ['rate', 'fdtr', 'target', 'estr', 'shibor'],
    'Surprise': ['surprise', 'cesi'],
}

def tokenize_words(name: str) -> set[str]:
    cleaned = ''.join(ch.lower() if ch.isalnum() else ' ' for ch in name)
    return {token for token in cleaned.split() if token}

def classify_asset(name: str) -> str:
    lname = name.lower()
    tokens = tokenize_words(name)
    if any(keyword in lname for keyword in macro_indicator_keywords):
        return 'Macro Indicator'
    for label, keywords in asset_rules:
        for keyword in keywords:
            if ' ' in keyword:
                if keyword in lname:
                    return label
            else:
                if keyword in tokens:
                    return label
    return 'Other Asset'

def classify_macro_bucket(name: str) -> str:
    lname = name.lower()
    for bucket, keywords in macro_keywords.items():
        if any(keyword in lname for keyword in keywords):
            return bucket
    return 'Other Macro'

series_metadata_rows = []
for series_name in aligned_panel.columns:
    series_metadata_rows.append({
        'series': series_name,
        'asset_class': classify_asset(series_name),
        'macro_bucket': classify_macro_bucket(series_name),
    })
series_metadata = pd.DataFrame(series_metadata_rows).set_index('series')

returns_monthly = returns.resample('ME').sum(min_count=1)
diffs_monthly = diffs.resample('ME').mean()

asset_group_returns = {}
for asset_class in sorted(series_metadata['asset_class'].unique()):
    if asset_class in {'Other Asset', 'Macro Indicator'}:
        continue
    cols = series_metadata.index[series_metadata['asset_class'] == asset_class]
    if len(cols) == 0:
        continue
    asset_group_returns[asset_class] = returns_monthly[cols].mean(axis=1)
asset_group_df = pd.DataFrame(asset_group_returns).dropna(how='all')

macro_group_changes = {}
for macro_bucket in sorted(series_metadata['macro_bucket'].unique()):
    if macro_bucket == 'Other Macro':
        continue
    cols = series_metadata.index[series_metadata['macro_bucket'] == macro_bucket]
    shared_cols = [col for col in cols if col in diffs_monthly.columns]
    if not shared_cols:
        continue
    agg = diffs_monthly[shared_cols].mean(axis=1)
    macro_group_changes[macro_bucket] = zscore_normalize(agg)
macro_factors_df = pd.DataFrame(macro_group_changes).dropna(how='all')
macro_factors_df = macro_factors_df.replace([np.inf, -np.inf], np.nan).dropna(how='all')

macro_beta_rows = []
macro_corr_rows = []
for asset_class, series in asset_group_df.items():
    aligned = pd.concat([series.rename('target'), macro_factors_df], axis=1, join='inner').dropna()
    if aligned.empty:
        continue
    X = aligned.iloc[:, 1:].to_numpy()
    y = aligned['target'].to_numpy()
    X = np.column_stack([np.ones(len(X)), X])
    beta, *_ = np.linalg.lstsq(X, y, rcond=None)
    fitted = X @ beta
    resid = y - fitted
    denom = ((y - y.mean()) ** 2).sum()
    r_squared = 1.0 - (resid ** 2).sum() / denom if denom != 0 else np.nan
    for factor, coefficient in zip(aligned.columns[1:], beta[1:]):
        macro_beta_rows.append({
            'asset_class': asset_class,
            'macro_factor': factor,
            'beta': coefficient,
            'r_squared': r_squared,
        })
        corr_val = aligned['target'].corr(aligned[factor])
        macro_corr_rows.append({
            'asset_class': asset_class,
            'macro_factor': factor,
            'correlation': corr_val,
        })
macro_beta_table = pd.DataFrame(macro_beta_rows)
macro_corr_table = pd.DataFrame(macro_corr_rows)

market_dynamics = {
    'series_metadata': series_metadata,
    'asset_group_returns_monthly': asset_group_df,
    'macro_factors': macro_factors_df,
    'macro_beta_table': macro_beta_table,
    'macro_corr_table': macro_corr_table,
}

macro_beta_table.sort_values(['asset_class', 'beta'], ascending=[True, False]).head(10)



asset_class  macro_factor      beta  r_squared
3  Alternative  Policy Rates  0.005262   0.662374
1  Alternative     Inflation  0.001441   0.662374
4  Alternative      Surprise  0.000281   0.662374
2  Alternative     Liquidity  0.000248   0.662374
0  Alternative        Growth -0.000867   0.662374
8  Commodities  Policy Rates  0.017883   0.330184
6  Commodities     Inflation  0.006873   0.330184
7  Commodities     Liquidity -0.000456   0.330184
5  Commodities        Growth -0.000797   0.330184
9  Commodities      Surprise -0.007414   0.330184

## 9.1 Macro Impact Interpretation
The regression table highlights which macro composites load most strongly on each asset bucket. Large positive betas signal assets that benefit when the factor rises, while negative betas expose sensitivity to deteriorating macro prints. Check the `r_squared` to gauge how much of the asset group variance each factor set explains and use the correlation table for a quick sanity check on directionality.


In [118]:
asset_group_df = market_dynamics['asset_group_returns_monthly']
if asset_group_df.empty:
    raise ValueError('No asset group returns available for cross-asset analysis.')

lags = range(1, 6)
lead_lag_records = []
for src in asset_group_df.columns:
    for dst in asset_group_df.columns:
        if src == dst:
            continue
        base = pd.concat([asset_group_df[src], asset_group_df[dst]], axis=1, join='inner').dropna()
        if base.empty:
            continue
        for lag in lags:
            corr_val = base[src].shift(-lag).corr(base[dst])
            if pd.notna(corr_val):
                lead_lag_records.append({
                    'leader': src,
                    'follower': dst,
                    'lag_months': lag,
                    'lead_correlation': corr_val,
                })
lead_lag_df = pd.DataFrame(lead_lag_records)

if not lead_lag_df.empty:
    top_leads = lead_lag_df.sort_values('lead_correlation', ascending=False).head(10)
else:
    top_leads = pd.DataFrame()

market_dynamics['lead_lag_table'] = lead_lag_df
market_dynamics['top_leads'] = top_leads

top_leads


leader     follower  lag_months  lead_correlation
250   Volatility  Commodities           1          0.346661
13   Alternative       Crypto           4          0.342233
215        Rates  Commodities           1          0.239470
48   Commodities       Crypto           4          0.226029
245   Volatility  Alternative           1          0.224870
216        Rates  Commodities           2          0.219031
106       Crypto  Alternative           2          0.213564
122       Crypto       Equity           3          0.190918
183           FX  Commodities           4          0.190345
112       Crypto  Commodities           3          0.180227

## 9.2 Cross-Asset Lead/Lag Insights
The lead-lag table lists which asset buckets historically move ahead of others over 1–5 month horizons. High positive `lead_correlation` values suggest the leader offers an early signal for the follower. Use this to spot propagation chains (e.g., rates leading equities) and focus on consistent leaders for tactical allocation.


In [119]:
regimes = nexus_index_bundle['regimes']
returns = nexus_index_bundle['stationary']['returns']
series_metadata = market_dynamics['series_metadata']

regime_distribution = regimes.apply(lambda col: col.value_counts(normalize=True)).T.fillna(0.0)

regime_performance_rows = []
for series_name in regimes.columns:
    if series_name not in returns.columns:
        continue
    joined = pd.DataFrame({
        'return': returns[series_name],
        'regime': regimes[series_name],
    }).dropna()
    if joined.empty:
        continue
    grouped = joined.groupby('regime')['return'].mean()
    for regime_label, avg_ret in grouped.items():
        regime_performance_rows.append({
            'series': series_name,
            'asset_class': series_metadata.loc[series_name, 'asset_class'],
            'regime': regime_label,
            'avg_return': avg_ret,
        })
regime_performance = pd.DataFrame(regime_performance_rows)

market_dynamics['regime_distribution'] = regime_distribution
market_dynamics['regime_performance'] = regime_performance

regime_distribution.head()


bear      bull  \
BBG EA Corp Avg OAS - LECPOAS INDEX Index           0.000472  0.000000   
BBG PE HY Avg OAS - LP01OAS INDEX Index             0.232503  0.152330   
BE 10 Year - USGGBE10 Index Index                   0.216846  0.240709   
Bloomberg Galaxy Crypto Index - BGCI Index Index    0.089794  0.091869   
Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index ...  0.041219  0.035559   

                                                    stagnant  
BBG EA Corp Avg OAS - LECPOAS INDEX Index           0.999528  
BBG PE HY Avg OAS - LP01OAS INDEX Index             0.615167  
BE 10 Year - USGGBE10 Index Index                   0.542445  
Bloomberg Galaxy Crypto Index - BGCI Index Index    0.818336  
Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index ...  0.923222

## 9.3 Regime Diagnostics Reading
`regime_distribution` shows the percentage of time each market spends in bull, bear, or stagnant states, while `regime_performance` reports the average return delivered in those regimes. Concentrate on series with skewed regime exposure or asymmetric payoffs to refine hedging and gear portfolio tilts toward favourable phases.


In [120]:
features = nexus_index_bundle['features']
anomalies = nexus_index_bundle['anomalies']
returns = nexus_index_bundle['stationary']['returns']

behaviour_rows = []
if isinstance(features.columns, pd.MultiIndex) and 'rsi14' in features.columns.get_level_values(1):
    rsi_panel = features.xs('rsi14', level=1, axis=1)
    next_returns = returns.shift(-5)
    for series_name in rsi_panel.columns:
        rsi_series = rsi_panel[series_name].dropna()
        if rsi_series.empty:
            continue
        overbought = rsi_series[rsi_series > 70].index
        oversold = rsi_series[rsi_series < 30].index
        overbought_ret = next_returns.loc[overbought, series_name].mean()
        oversold_ret = next_returns.loc[oversold, series_name].mean()
        behaviour_rows.append({
            'series': series_name,
            'signal': 'RSI>70 forward 5d',
            'avg_forward_return': overbought_ret,
            'signal_count': len(overbought),
        })
        behaviour_rows.append({
            'series': series_name,
            'signal': 'RSI<30 forward 5d',
            'avg_forward_return': oversold_ret,
            'signal_count': len(oversold),
        })

if anomalies is not None:
    for series_name in anomalies.columns:
        flagged = anomalies[series_name] == 1
        if flagged.any() and series_name in returns.columns:
            subsequent = returns[series_name].shift(-1)
            behaviour_rows.append({
                'series': series_name,
                'signal': 'Anomaly flag next day',
                'avg_forward_return': subsequent[flagged].mean(),
                'signal_count': int(flagged.sum()),
            })

behaviour_table = pd.DataFrame(behaviour_rows)
market_dynamics['behaviour_table'] = behaviour_table

behaviour_table.head(10)


series             signal  \
0          BBG EA Corp Avg OAS - LECPOAS INDEX Index  RSI>70 forward 5d   
1          BBG EA Corp Avg OAS - LECPOAS INDEX Index  RSI<30 forward 5d   
2            BBG PE HY Avg OAS - LP01OAS INDEX Index  RSI>70 forward 5d   
3            BBG PE HY Avg OAS - LP01OAS INDEX Index  RSI<30 forward 5d   
4                  BE 10 Year - USGGBE10 Index Index  RSI>70 forward 5d   
5                  BE 10 Year - USGGBE10 Index Index  RSI<30 forward 5d   
6   Bloomberg Galaxy Crypto Index - BGCI Index Index  RSI>70 forward 5d   
7   Bloomberg Galaxy Crypto Index - BGCI Index Index  RSI<30 forward 5d   
8  Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index...  RSI>70 forward 5d   
9  Bloomberg Nominal USD 5Y Spot - BTSIUS05 Index...  RSI<30 forward 5d   

   avg_forward_return  signal_count  
0                 NaN             0  
1                 NaN             5  
2            0.001740           376  
3           -0.000878           723  
4            0.000130           265  
5           -0.000823           245  
6            0.001468           195  
7            0.004952            96  
8            0.000675            15  
9           -0.000045            71

## 9.4 Behavioural Signal Takeaways
RSI-based rows summarise forward performance after classic overbought/oversold episodes; persistent negative follow-through on `RSI>70` hints at mean-reverting behaviour. The anomaly flag statistics quantify how price spikes or volatility shocks resolve—use them to calibrate stop-loss rules or contrarian entries.


In [121]:
summary = {
    'macro_factors_available': list(market_dynamics['macro_factors'].columns),
    'asset_groups': list(market_dynamics['asset_group_returns_monthly'].columns),
    'strongest_macro_link': market_dynamics['macro_beta_table'].sort_values('beta', ascending=False).head(3)[['asset_class', 'macro_factor', 'beta', 'r_squared']].to_dict('records'),
    'top_leads': market_dynamics.get('top_leads', pd.DataFrame()).head(5).to_dict('records'),
}
summary


{'macro_factors_available': ['Growth',
  'Inflation',
  'Liquidity',
  'Policy Rates',
  'Surprise'],
 'asset_groups': ['Alternative',
  'Commodities',
  'Credit',
  'Crypto',
  'Equity',
  'FX',
  'Rates',
  'Volatility'],
 'strongest_macro_link': [{'asset_class': 'Crypto',
   'macro_factor': 'Inflation',
   'beta': 0.044012531036911814,
   'r_squared': 0.19906851985932938},
  {'asset_class': 'Crypto',
   'macro_factor': 'Surprise',
   'beta': 0.042336658073801284,
   'r_squared': 0.19906851985932938},
  {'asset_class': 'Volatility',
   'macro_factor': 'Surprise',
   'beta': 0.03849855868327081,
   'r_squared': 0.267544115550994}],
 'top_leads': [{'leader': 'Volatility',
   'follower': 'Commodities',
   'lag_months': 1,
   'lead_correlation': 0.34666148561295795},
  {'leader': 'Alternative',
   'follower': 'Crypto',
   'lag_months': 4,
   'lead_correlation': 0.34223319880453323},
  {'leader': 'Rates',
   'follower': 'Commodities',
   'lag_months': 1,
   'lead_correlation': 0.239470435

## 9.5 Market Dynamics Summary
This dictionary provides a quick digest of the computed artefacts—verify macro factors and asset groups aligned with your universe, inspect the strongest macro linkage, and review top lead relationships before drilling into the detailed tables for trade ideas.


# 10. Market Dynamics — Detailed Report
This section generates a detailed narrative using the computed Market Dynamics artefacts.
Run the preprocessing and Market Dynamics blocks first (stages 1–10 and Market Dynamics cells).


In [ ]:

from IPython.display import Markdown, display
import pandas as pd
import numpy as np

# Guards and helpers
if 'nexus_index_bundle' not in globals():
    raise RuntimeError("Please run the preprocessing pipeline to create 'nexus_index_bundle' before generating the report.")

if 'market_dynamics' not in globals():
    raise RuntimeError("Please run the Market Dynamics cells to populate 'market_dynamics' before generating the report.")

bundle = nexus_index_bundle
md = market_dynamics

# Pull artefacts with safe fallbacks
asset_groups = list(md.get('asset_group_returns_monthly', pd.DataFrame()).columns)
macro_factors = list(md.get('macro_factors', pd.DataFrame()).columns)
macro_beta = md.get('macro_beta_table', pd.DataFrame())
macro_corr = md.get('macro_corr_table', pd.DataFrame())
lead_lag = md.get('lead_lag_table', pd.DataFrame())
top_leads = md.get('top_leads', pd.DataFrame())
regime_distribution = md.get('regime_distribution', pd.DataFrame())
regime_performance = md.get('regime_performance', pd.DataFrame())
behaviour = md.get('behaviour_table', pd.DataFrame())

# Build summaries
lines = []
lines.append("**Executive Summary**")
lines.append(f"- Asset groups: {', '.join(asset_groups) if asset_groups else 'n/a'}")
lines.append(f"- Macro composites: {', '.join(macro_factors) if macro_factors else 'n/a'}")

# Macro linkage (top per asset class)
macro_lines = []
macro_top_tbl = pd.DataFrame()
if not macro_beta.empty:
    macro_top_tbl = (
        macro_beta.sort_values(['asset_class','beta'], ascending=[True, False])
                  .groupby('asset_class', as_index=False)
                  .head(1)
                  .reset_index(drop=True)
    )
    for _, r in macro_top_tbl.iterrows():
        macro_lines.append(f"- {r['asset_class']}: {r['macro_factor']} beta={r['beta']:.3f}, R^2={r['r_squared']:.3f}")
else:
    macro_lines.append("- Macro beta table unavailable.")

# Correlation sanity checks (top 5 by absolute corr)
macro_corr_lines = []
macro_corr_top = pd.DataFrame()
if not macro_corr.empty:
    macro_corr_top = macro_corr.assign(abs_corr=macro_corr['correlation'].abs()).sort_values('abs_corr', ascending=False).head(5)
    for _, r in macro_corr_top.iterrows():
        macro_corr_lines.append(f"- {r['asset_class']} vs {r['macro_factor']}: corr={r['correlation']:.3f}")
else:
    macro_corr_lines.append("- Macro correlation table unavailable.")

# Lead/Lag
lead_lines = []
lead_tbl = pd.DataFrame()
if not top_leads.empty:
    lead_tbl = top_leads.head(10).copy()
    for _, r in lead_tbl.iterrows():
        lead_lines.append(f"- {r['leader']} -> {r['follower']} ({int(r['lag_months'])}m): corr={r['lead_correlation']:.3f}")
elif not lead_lag.empty:
    lead_tbl = lead_lag.sort_values('lead_correlation', ascending=False).head(10).copy()
    for _, r in lead_tbl.iterrows():
        lead_lines.append(f"- {r['leader']} -> {r['follower']} ({int(r['lag_months'])}m): corr={r['lead_correlation']:.3f}")
else:
    lead_lines.append("- Lead/lag table unavailable.")

# Regime diagnostics
regime_lines = []
regime_mix = pd.Series(dtype=float)
if not regime_distribution.empty:
    # Average distribution across series to get an equal-weighted mix
    regime_mix = regime_distribution.mean().sort_values(ascending=False)
    regime_lines.append("- Portfolio regime mix (equal-weighted across series):")
    for k, v in regime_mix.items():
        regime_lines.append(f"  - {k}: {v:.1%}")
else:
    regime_lines.append("- Regime distribution unavailable.")

# Behavioural (RSI) summary
beh_lines = []
beh_summary = pd.DataFrame()
if not behaviour.empty:
    beh_summary = behaviour.groupby('signal')['avg_forward_return'].mean().to_frame('avg_fwd').reset_index()
    for _, r in beh_summary.iterrows():
        beh_lines.append(f"- {r['signal']}: {r['avg_fwd']:.4%}")
else:
    beh_lines.append("- Behavioural table unavailable.")

# Recent monthly performance by asset group (last 3 months)
perf_lines = []
asset_monthly = md.get('asset_group_returns_monthly', pd.DataFrame())
if not asset_monthly.empty and asset_monthly.shape[0] >= 3:
    recent = asset_monthly.tail(3)
    mlabels = [d.strftime('%Y-%m') for d in recent.index]
    perf_lines.append("- Recent asset-group monthly returns (last 3 months):")
    for ac in asset_monthly.columns:
        vals = ''.join([f" {lab}:{val:.2%}" for lab, val in zip(mlabels, recent[ac])])
        perf_lines.append(f"  - {ac}:{vals}")

# Assemble Markdown report
report_sections = []
report_sections.append("\n".join(lines))
report_sections.append("**Macro Linkage**\n" + ("\n".join(macro_lines)))
report_sections.append("**Macro Correlations (Top 5)**\n" + ("\n".join(macro_corr_lines)))
report_sections.append("**Cross-Asset Lead/Lag (Top 10)**\n" + ("\n".join(lead_lines)))
report_sections.append("**Regime Diagnostics**\n" + ("\n".join(regime_lines)))
report_sections.append("**Behavioral Signals (RSI, fwd 5d)**\n" + ("\n".join(beh_lines)))
if perf_lines:
    report_sections.append("**Recent Monthly Performance**\n" + ("\n".join(perf_lines)))

full_report = "\n\n".join(report_sections)
display(Markdown(full_report))

# Optionally also show the underlying small tables for transparency
print("\n\nTables shown below for reference (top links, leads, regime mix, behavioral summary):")
if not macro_top_tbl.empty:
    display(macro_top_tbl[['asset_class','macro_factor','beta','r_squared']])
if not lead_tbl.empty:
    display(lead_tbl)
if not regime_mix.empty:
    display(regime_mix.to_frame('share'))
if not beh_summary.empty:
    display(beh_summary)



## 10.1 Executive Summary
- Asset groups: Alternative, Commodities, Credit, Crypto, Equity, FX, Rates, Volatility
- Macro composites: Growth, Inflation, Liquidity, Policy Rates, Surprise
- Dominant driver: Policy Rates loads strongest across several groups; Inflation salient for Crypto and FX.
- Lead–lag: Volatility and Rates tend to lead Commodities over 1–2 months; Crypto exhibits both leader/follower behavior across pairs.
- Regimes: Equal-weighted mix skewed to stagnant (>60%), with smaller bull and bear shares.
- Behavioral (RSI): Mild mean reversion — oversold bounces, overbought softens on average over the next 5 trading days.

## 10.2 Macro Linkage
### 10.2.1 Top Factor per Asset Class
- Alternative: Policy Rates beta=0.005, R^2=0.662
- Commodities: Policy Rates beta=0.018, R^2=0.330
- Credit: Surprise beta=0.001, R^2=0.213
- Crypto: Inflation beta=0.044, R^2=0.199
- Equity: Policy Rates beta=0.004, R^2=0.076
- FX: Inflation beta≈0.000, R^2=0.074
- Rates: Policy Rates beta=0.035, R^2=0.206
- Volatility: Surprise beta=0.038, R^2=0.268

### 10.2.2 Interpretation
- Policy Rate shocks broadly influence Alternatives, Commodities, and Rates; Equities show weaker but consistent sensitivity.
- Inflation factor has notable explanatory power for Crypto and FX, consistent with price-level and purchasing-power channels.
- Surprise effects (data-release shocks) matter for Credit and Volatility, aligning with risk-premia repricing behavior.

## 10.3 Cross‑Asset Lead/Lag
### 10.3.1 Top Relationships
- Volatility → Commodities (1m): corr=0.347
- Alternative → Crypto (4m): corr=0.342
- Rates → Commodities (1–2m): corr=0.239 to 0.219
- Commodities → Crypto (3–4m): corr=0.226 to 0.180
- Volatility → Alternative (1m): corr=0.225
- Crypto → Alternative (2m): corr=0.214
- Crypto → Equity (3m): corr=0.191
- FX → Commodities (4m): corr=0.190

### 10.3.2 Interpretation
- Volatility spikes often propagate into Commodities and Alternatives within ~1 month.
- Rates lead Commodities, consistent with macro expectations flowing into real assets.
- Crypto sits in a reflexive risk-on/off loop, sometimes leading Alternatives and Equities with multi‑month lags.

## 10.4 Regime Diagnostics
### 10.4.1 Portfolio Regime Mix (equal‑weighted)
- Stagnant: 65.9%
- Bull: 17.9%
- Bear: 16.2%

### 10.4.2 Interpretation
- A majority “stagnant” share indicates extended range‑bound conditions; tilts favor carry and mean‑reversion, with tempered trend following.

## 10.5 Behavioral Signals (RSI)
### 10.5.1 Forward 5 Trading Days
- RSI<30: +0.116% average forward return
- RSI>70: −0.070% average forward return

### 10.5.2 Interpretation
- Classical RSI thresholds are mildly contrarian on average — consider position sizing and risk filters; effects are small and noisy in isolation.

## 10.6 Recent Monthly Performance (Last 3 Months)
- Refer to the Market Dynamics Summary cell for exact month-by-month values; groups covered: Alternative, Commodities, Credit, Crypto, Equity, FX, Rates, Volatility.

## 10.7 Methodological Notes
- Monthly aggregation uses calendar month end (ME).
- Asset grouping excludes series classified as “Macro Indicator” to avoid contaminating return groups with economic levels.
- Macro composites are z‑scored averages of series within macro buckets.
- Macro linkage uses simple OLS betas and R^2 on monthly aggregates; lead/lag is based on shifted correlations across 1–5 month horizons.

## 10.8 Next Steps
- Stress test macro betas by sub‑periods and rolling windows.
- Validate lead/lag stability via bootstrapping and out‑of‑sample splits.
- Combine RSI with regime filters to improve signal quality and risk‑adjusted returns.
- Extend macro buckets (e.g., financial conditions, breakeven terms) and re‑evaluate exposures.
